src/main/java/core/basesyntax/ShopServiceImpl.java

```
@@ -0,0 +1,31 @@
+package core.basesyntax;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class ShopServiceImpl implements ShopService {
+    private final OperationStrategy operationStrategy;
+    private final Map<String, Integer> storage = new HashMap<>();
+
+    public ShopServiceImpl(OperationStrategy operationStrategy) {
+        this.operationStrategy = operationStrategy;
+    }
+
+    @Override
+    public void process(List<FruitTransaction> transactions) {
+        for (FruitTransaction transaction : transactions) {
+            OperationHandler handler = operationStrategy.getHandler(transaction.getOperation());
+            if (handler != null) {
+                handler.handle(storage, transaction);
+            } else {
+                System.err.println("No handler found for operation: " + transaction.getOperation());
+            }
```

The responsibility for finding a handler should lie entirely within the `OperationStrategy`. Instead of checking for `null` here and printing an error, the `operationStrategy.getHandler()` method should throw an exception (e.g., `IllegalStateException`) if no handler is found. This "fail-fast" approach makes the system more robust, as it prevents the application from continuing in an invalid state. This service can then simply call `handler.handle(...)` without the null check.